internal/handlers/common/aggregations/group.go

```diff
@@ -0,0 +1,319 @@
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+)
+
+// newAccumulatorFunc is a type for a function that creates an accumulator.
+type newAccumulatorFunc func(expression *types.Document) (Accumulator, error)
+
+// Accumulator is a common interface for accumulation.
+type Accumulator interface {
+	// Accumulate documents and returns the result of accumulation.
+	Accumulate(ctx context.Context, in []*types.Document) (any, error)
+}
+
+// accumulators maps all supported $group accumulators.
+var accumulators = map[string]newAccumulatorFunc{
+	// sorted alphabetically
+	"$count": newCountAccumulator,
+}
+
+// groupStage represents $group stage.
+//
+//	{ $group: {
+//		_id: <groupExpression>,
+//		<groupBy[0].outputField>: {accumulator0: expression0},
+//		...
+//		<groupBy[N].outputField>: {accumulatorN: expressionN},
+//	}}
+type groupStage struct {
+	groupExpression any
+	groupBy         []groupBy
+}
+
+// groupBy represents accumulation to apply on the group.
+type groupBy struct {
+	accumulate  func(ctx context.Context, in []*types.Document) (any, error)
+	outputField string
+}
+
+// newGroup creates a new $group stage.
+func newGroup(stage *types.Document) (Stage, error) {
+	fields, err := common.GetRequiredParam[*types.Document](stage, "$group")
+	if err != nil {
+		return nil, commonerrors.NewCommandErrorMsgWithArgument(
+			commonerrors.ErrStageGroupInvalidFields,
+			"a group's fields must be specified in an object",
+			"$group",
+		)
+	}
+
+	var groupKey any
+	var groups []groupBy
+
+	iter := fields.Iterator()
+
+	defer iter.Close()
+
+	for {
+		field, v, err := iter.Next()
+		if errors.Is(err, iterator.ErrIteratorDone) {
+			break
+		}
+
+		if err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		if field == "_id" {
+			groupKey = v
+
+			idAccumulator := idAccumulator{
+				expression: v,
+			}
+			groups = append(groups, groupBy{
+				outputField: field,
+				accumulate:  idAccumulator.Accumulate,
+			})
+
+			continue
+		}
+
+		accumulation, ok := v.(*types.Document)
+		if !ok || accumulation.Len() == 0 {
+			return nil, commonerrors.NewCommandErrorMsgWithArgument(
+				commonerrors.ErrStageGroupInvalidAccumulator,
+				fmt.Sprintf("The field '%s' must be an accumulator object", field),
+				"$group",
+			)
+		}
+
+		// accumulation document contains only one field.
+		if accumulation.Len() > 1 {
+			return nil, commonerrors.NewCommandErrorMsgWithArgument(
+				commonerrors.ErrStageGroupMultipleAccumulator,
+				fmt.Sprintf("The field '%s' must specify one accumulator", field),
+				"$group",
+			)
+		}
+
+		operator, _, err := accumulation.Iterator().Next()
+		if err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		newAccumulator, ok := accumulators[operator]
+		if !ok {
+			return nil, commonerrors.NewCommandErrorMsgWithArgument(
+				commonerrors.ErrNotImplemented,
+				"unimplemented",
```

Comment:
That's a very bad error message that is confusing for real users: https://ferretdb.slack.com/archives/C049W6E01E0/p1678723727614249?thread_ts=1678704507.197499&cid=C049W6E01E0 The message should describe what is not implemented.